Silicon/TigerlakePkg/Library/TxtInit/x64/TxtDxeBsp64.nasm

```diff
@@ -0,0 +1,849 @@
+
+    ;
+    ; Re-enable paging
+    ;
+    mov     rbx, cr0
+    bts     ebx, 31
+    mov     cr0, rbx
+
+    ;
+    ; Now we're in Compatibility mode
+    ;
+    MOVQOUT3                        ; rax = the address of our local variable
+    lgdt    [rax]
+    lidt    [rax-0x10]
+
+    ;
+    ; Reload cs register
+    ;
+    retf
+LaunchBiosAcm_ReloadCS:
+    ;
+    ; Now we're in Long Mode
+    ;
+
+    ;
+    ; Restore the rest of the segment registers
+    ;
+    pop     rax
+    mov     ds, ax
+    pop     rax
+    mov     es, ax
+    pop     rax
+    mov     fs, ax
+    pop     rax
+    mov     gs, ax
+
+;-----------------------------------------------------------------------------
+;
+;  Section:  Restore MTRRs
+;
+;  Description:  BIOS MTRR values are restored.
+;
+;-----------------------------------------------------------------------------
+    ;
+    ; Clear out the cache
+    ;
+    mov     rax, cr0                ; Set CR0:CD and CR0:NE, clear CR0:NW
+    or      rax, CR0_CD_MASK | CR0_NE_MASK
+    and     rax, ~CR0_NW_MASK
+    mov     cr0, rax
+    wbinvd                          ; Flush and invalidate the cache
+
+    ;
+    ; Restore def MTRR type
+    ;
+    mov     rcx, IA32_MTRR_DEF_TYPE
+    pop     rdx
+    pop     rax
+    wrmsr
+
+    ;
+    ; Reset Variable MTRRs to Pre-GETSEC state
+    ;
+    mov     rcx, IA32_MTRR_CAP
+    rdmsr
+    and     rax, 0xFF
+    shl     rax, 1
+    mov     rcx, rax
+    mov     rbx, rax
+    shl     rbx, 4                  ; rbx = the total length for MTRR register
+    add     rsp, rbx                ; adjust rsp to skip MTRR register
+    mov     rbx, rsp
+    sub     rbx, 8                  ; let rbx point to head of MTRR register now
+
+LaunchBiosAcm_RestoreNextMtrr:
+    add     rcx, IA32_MTRR_PHYSBASE0 - 1
+    mov     rdx, [rbx]
+    mov     rax, [rbx - 8]
+    wrmsr
+    sub     rcx, IA32_MTRR_PHYSBASE0 - 1
+    sub     rbx, 16
+    loop    LaunchBiosAcm_RestoreNextMtrr
+
+    ;
+    ; Restore IA32_EFER MSR
+    ;
+    pop     rdx
+    pop     rax
+    mov     rcx, IA32_EFER_MSR
+    wrmsr
+
+    ;
+    ; Restore MISC ENABLE MSR
+    ;
+    pop     rdx
+    pop     rax
+    mov     rcx, IA32_MISC_ENABLE
+    wrmsr
+
+;-----------------------------------------------------------------------------
+;
+;  Section:      Restore registers, stack and exit.
+;
+;  Description:  Previously saved registers are restored. Stack is restored
+;                by execution leave instruction and control is returned to
+;                caller.
+;
+;         NOTE!
+;         This section ends differences between programmed and executed code.
+;
+;-----------------------------------------------------------------------------
+    ;
+    ; Now we're in Long Mode
+    ; Restore control registers
+    ;
+    pop     rax
+    mov     cr0, rax
+    pop     rax
+    mov     cr4, rax
+
+    wbinvd                          ; Flush and invalidate the cache
+
+LaunchBiosAcm_SkipLaunch:
+    POPA_64
+    popfq
```

Comment:
Does 64 bit is validated to TGL?

Reply:
64 bits is not tested. Infact tgl_bios does not have ia23 code anymore. should sbl only have x64 also?